### PR TITLE
Tickets search and choice

### PR DESCRIPTION
**Pull request by ArinaSavchenko**

### CODE REVIEW

File: Airline_Web_API/Airlina_DB/dbo/Tables/TicketTypes.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿CREATE TABLE [dbo].[TicketTypes]
+(
+	[TicketTypeId] INT IDENTITY(1,1) NOT NULL PRIMARY KEY,
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airlina_DB/dbo/Tables/TicketTypes.sql]: As this field is inside TicketType table you don't need to duplicate this information in field name, So just [Id] is enough.

File: Airline_Web_API/Airlina_DB/dbo/Tables/Tickets.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿CREATE TABLE [dbo].[Tickets]
+(
+	[TicketId] INT IDENTITY(1,1) NOT NULL PRIMARY KEY,
+	[FlightId] INT NOT NULL,
+	[TicketTypeId] INT NOT NULL,
+	[Price] MONEY NOT NULL,
+	[TotalNumber] INT NOT NUll,
+	[TicketsLeft] INT NOT NULL,
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airlina_DB/dbo/Tables/Tickets.sql]: Please review naming of these columns. They are both represent number of tickets but named differently.

File: Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs [hunk 3]
```diff
@@ -22,17 +22,26 @@ public FlightsController(AirlineContext context)
 
         // GET: api/flights
         [HttpGet]
-        public async Task<ActionResult<IEnumerable<Flight>>> GetFlights([FromQuery] int departure, int arrival, DateTime date)
+        public async Task<ActionResult<IEnumerable<Flight>>> GetFlights([FromQuery] int departure, int arrival, DateTime date, int amount)
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs]: Please review naming. 'amount' of what exactly? Also arrival and departure with type int, what are they?

File: Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs [hunk 4]
```diff
@@ -22,17 +22,26 @@ public FlightsController(AirlineContext context)
...
+                .Include(flight => flight.ArrivalAirport)
+                .Include(flight => flight.Tickets)
+                .AsQueryable();
+            flights = flights.Where(flight =>
+                flight.DepartureAirportId == departure
+                && flight.ArrivalAirportId == arrival
+                && flight.DepartureDate.Date == date.Date
+                && flight.Tickets
+                    .Where(ticket => ticket.TicketsLeft >= amount)
+                    .Count() > 0
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs]: You can combine these expressions into one Count()

File: Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs [hunk 5]
```diff
@@ -22,17 +22,26 @@ public FlightsController(AirlineContext context)
...
             );
+                
             return await flights
                 .OrderBy(flight => flight.FlightId)
                 .ToListAsync();
+
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs:48]: Unnecessary empty line.

File: Airline_Web_API/Airline_Web_API/Startup.cs [hunk 6]
```diff
@@ -29,6 +29,10 @@ public void ConfigureServices(IServiceCollection services)
             services.AddDbContext<AirlineContext>(options
                => options.UseSqlServer(Configuration.GetConnectionString("AirlineDatabase")));
             services.AddControllers();
+            services.AddControllersWithViews()
+                    .AddNewtonsoftJson(options =>
+                    options.SerializerSettings.ReferenceLoopHandling = Newtonsoft.Json.ReferenceLoopHandling.Ignore
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Startup.cs]: Why do you need this?

File: ClientApp/src/app/FlightForSearch.ts [hunk 7]
```diff
@@ -2,6 +2,5 @@ export interface FlightForSearch{
   departure?: number;
   arrival?: number;
   date?: any;
```
Comment (andrei-kruhliakou-itechart) [inline ClientApp/src/app/FlightForSearch.ts]: Why does this field have type 'any'?

File: ClientApp/src/app/Ticket.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+export interface Ticket {
+  // id: number;
+  // flightId: number;
+  // ticketTypeId:
```
Comment (andrei-kruhliakou-itechart) [inline ClientApp/src/app/Ticket.ts]: Please do not commit commented out code

**Review by andrei-kruhliakou-itechart — CHANGES_REQUESTED** [review]

File: Airline_Web_API/Airline_Web_API/Startup.cs (hunk 6 above)
Comment (ArinaSavchenko) [reply Airline_Web_API/Airline_Web_API/Startup.cs]: I need it to allow converting tickets array and airports in returned flight to JSON

File: Airline_Web_API/Airlina_DB/dbo/Tables/Tickets.sql [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿CREATE TABLE [dbo].[Tickets]
+(
+	[Id] INT IDENTITY(1,1) NOT NULL PRIMARY KEY,
+	[FlightId] INT NOT NULL,
+	[TicketTypeId] INT NOT NULL,
+	[Price] MONEY NOT NULL,
+	[TotalNumber] INT NOT NUll,
+	[TicketsLeftNumber] INT NOT NULL,
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airlina_DB/dbo/Tables/Tickets.sql:8]: This field is in Tickets table, why should it have Tickets prefix?

**Review by andrei-kruhliakou-itechart — CHANGES_REQUESTED** [review]

File: Airline_Web_API/Airline_Web_API/Controllers/AirportsController.cs [hunk 10]
```diff
@@ -32,9 +34,18 @@ public async Task<ActionResult<IEnumerable<Airport>>> GetAirports([FromQuery] st
...
-                .ToListAsync();
+
+            var airportsSearchResults = await airports.OrderBy(airport => airport.Id).ToListAsync();
+
+            var config = new MapperConfiguration(config =>
+            {
+                config.CreateMap<Airport, AirportViewModel>();
+            });
+
+            var mapper = new Mapper(config);
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/AirportsController.cs]: As this does not change between requests you can move it to static constructor and use static mapper in code.

File: Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs [hunk 11]
```diff
@@ -28,18 +30,33 @@ public async Task<ActionResult<IEnumerable<Flight>>> GetFlights([FromQuery] int
...
+            var config = new MapperConfiguration(config =>
+            {
+                config.CreateMap<Airport, AirportViewModel>();
+                config.CreateMap<TicketType, TicketTypeViewModel>();
+                config.CreateMap<Ticket, TicketsViewModel>();
+                config.CreateMap<Flight, FlightViewModel>();
+            }
+           );
+
+            var mapper = new Mapper(config);
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs]: As this does not change between requests you can move it to static constructor and use static mapper in code.

File: Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs [hunk 12]
```diff
@@ -28,18 +30,33 @@ public async Task<ActionResult<IEnumerable<Flight>>> GetFlights([FromQuery] int
...
                 && flight.DepartureDate.Date == date.Date
                 && flight.Tickets
                     .Count(ticket => ticket.TicketsLeftNumber >= ticketsNumber) > 0
             );
-                
-            return await flights
-                .OrderBy(flight => flight.Id)
-                .ToListAsync();
+
+            var flightsSearchResults = await flights.OrderBy(flight => flight.Id).ToListAsync();
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs]: Every operation from new line please.

File: Airline_Web_API/Airline_Web_API/ViewModels/FlightViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    public class FlightViewModel
+    {
+        public int Id { get; set; }
+        public DateTime DepartureDate { get; set; }
+        public DateTime ArrivalDate { get; set; }
+
+        public virtual AirportViewModel DepartureAirport { get; set; }
+        public virtual AirportViewModel ArrivalAirport { get; set; }
+
+        public virtual ICollection<TicketsViewModel> Tickets { get; set; }
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/ViewModels/FlightViewModel.cs]: Do you need virtual here?

File: ClientApp/src/app/Models/Ticket.ts [hunk 14]
```diff
@@ -1,3 +1,7 @@
+import { TicketType } from './TicketType';
+
 export interface Ticket {
   id: number;
+  ticketType: TicketType;
+  price: any;
```
Comment (andrei-kruhliakou-itechart) [inline ClientApp/src/app/Models/Ticket.ts]: Why the price is of any type?

File: ClientApp/src/app/flight-search-results/flight-search-results.component.html [hunk 15]
```diff
@@ -10,4 +10,5 @@
       Flights BACK: {{flight.flightId}} {{flight.departureDate}} {{flight.departureAirport.country}}
     </li>
   </ul>
+  <>
```
Comment (andrei-kruhliakou-itechart) [inline ClientApp/src/app/flight-search-results/flight-search-results.component.html]: What's that for?

**Review by andrei-kruhliakou-itechart — CHANGES_REQUESTED** [review]

File: Airline_Web_API/Airline_Web_API/Controllers/AirportsController.cs [hunk 16]
```diff
@@ -35,7 +35,9 @@ public async Task<ActionResult<IEnumerable<Airport>>> GetAirports([FromQuery] st
                 airports = airportsByName.Union(airportsByCity);
             }
 
-            var airportsSearchResults = await airports.OrderBy(airport => airport.Id).ToListAsync();
+            var airportsSearchResults = await airports
+                                                .OrderBy(airport => airport.Id)
+                                                .ToListAsync();
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/AirportsController.cs]: Use one tab for indent please

File: Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs [hunk 17]
```diff
@@ -41,20 +41,11 @@ public async Task<ActionResult<IEnumerable<Flight>>> GetFlights([FromQuery] int
                     .Count(ticket => ticket.TicketsLeftNumber >= ticketsNumber) > 0
             );
 
-            var flightsSearchResults = await flights.OrderBy(flight => flight.Id).ToListAsync();
+            var flightsSearchResults = await flights
+                                                .OrderBy(flight => flight.Id)
+                                                .ToListAsync();
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Controllers/FlightsController.cs]: Use one tab for indent please

File: Airline_Web_API/Airline_Web_API/Models/AirlineContext.cs [hunk 18]
```diff
@@ -13,6 +15,21 @@ public class AirlineContext: DbContext
...
+            {
+                config.CreateMap<Airport, AirportViewModel>();
+                config.CreateMap<TicketType, TicketTypeViewModel>();
+                config.CreateMap<Ticket, TicketsViewModel>();
+                config.CreateMap<Flight, FlightViewModel>();
+            }
+           );
+          mapper = new Mapper(config);
+        }
+
```
Comment (andrei-kruhliakou-itechart) [inline Airline_Web_API/Airline_Web_API/Models/AirlineContext.cs]: Please move this to separate class

**Review by andrei-kruhliakou-itechart — CHANGES_REQUESTED** [review]

**Review by andrei-kruhliakou-itechart — APPROVED** [review]

**Pull request merged**